flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java

```diff
@@ -96,28 +97,115 @@ public JdbcDynamicTableSource(
+                 * We cannot use ParameterizedPredicate as it takes the filter expression as input (e.g EQUALS(...)
+                 * not the form we have here an example would be ('field1'= ?).
+                 *
+                 * An entry in the resolvedPredicates list may have more than one associated pushdown parameter, for example
+                 * a query like this : ... on e.type = 2 and (e.age = 50 OR height > 90)  and a.ip = e.ip;
+                 * will have 2 resolvedPredicates and 3 pushdownParams. The 2nd and 3rd pushdownParams will be for the second
+                 * resolvedPredicate.
+                 *
+                 */
+                ArrayList<String> paramsForThisPredicate = new ArrayList();
+                char placeholderChar =
+                        JdbcFilterPushdownPreparedStatementVisitor.PUSHDOWN_PREDICATE_PLACEHOLDER
+                                .charAt(0);
+
+                int count =
+                        (int) resolvedPredicate.chars().filter(ch -> ch == placeholderChar).count();
+
+                for (int j = processedPushdownParamsIndex;
+                        j < processedPushdownParamsIndex + count;
+                        j++) {
+                    paramsForThisPredicate.add(this.pushdownParams[j].toString());
+                }
+                processedPushdownParamsIndex = processedPushdownParamsIndex + count;
+
+                conditions[i] =
+                        resolvePredicateWithParams(
+                                resolvedPredicate, paramsForThisPredicate, placeholderChar);
+            }
+        }
+
         JdbcRowDataLookupFunction lookupFunction =
                 new JdbcRowDataLookupFunction(
                         options,
                         lookupMaxRetryTimes,
                         DataType.getFieldNames(physicalRowDataType).toArray(new String[0]),
                         DataType.getFieldDataTypes(physicalRowDataType).toArray(new DataType[0]),
                         keyNames,
-                        rowType);
+                        rowType,
+                        conditions);
         if (cache != null) {
             return PartialCachingLookupProvider.of(lookupFunction, cache);
         } else {
             return LookupFunctionProvider.of(lookupFunction);
         }
     }
 
+    /**
+     * Replace all instances of with the supplied params in order, as long as the characters are not
+     * in quotes.
+     *
+     * @param string string containing the placeholder characters
+     * @param params parameters to replace the the placeholder characters with.
+     * @return resolved predicate with parameters replacing the placeholder character
+     */
+    @VisibleForTesting
+    protected static String resolvePredicateWithParams(
+            String string, ArrayList<String> params, char placeholderChar) {
+        final char quoteIdentifier = '`';
+        StringBuffer resolvedPredicateWithParams = new StringBuffer();
```

Comment:
`StringBuffer` is thread safe meaning it has corresponding trade offs for that however we don't have multiple threads using this var, so I think `StringBuilder` should be enough